compiler/rustc_metadata/src/locator.rs

```diff
@@ -736,10 +718,12 @@ impl<'a> CrateLocator<'a> {
                 let loc_canon = loc.canonicalized().clone();
                 let loc = loc.original();
                 if loc.file_name().unwrap().to_str().unwrap().ends_with(".rlib") {
+                    rmetas.insert(loc_canon.with_extension("rmeta"), PathKind::ExternFlag);
                     rlibs.insert(loc_canon, PathKind::ExternFlag);
                 } else if loc.file_name().unwrap().to_str().unwrap().ends_with(".rmeta") {
                     rmetas.insert(loc_canon, PathKind::ExternFlag);
                 } else {
+                    rmetas.insert(loc_canon.with_extension("rmeta"), PathKind::ExternFlag);
```

Comment:
This is a hack. It should probably be replaced with proper lookup in the library search path, or with cargo passing `--extern` twice for both the rlib/dylib and the rmeta file.